include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,632 @@
+          const LinearOperator<Range, Domain> &op)
+{
+  LinearOperator<Range, Domain> return_op = op;
+
+  // ensure to have valid computation objects by catching number and op by
+  // value
+
+  return_op.vmult = [number, op](Range &v, const Domain &u)
+  {
+    op.vmult(v,u);
+    v *= number;
+  };
+
+  return_op.vmult_add = [number, op](Range &v, const Domain &u)
+  {
+    // FIXME: Is there a better way of doing this?
+    v /= number;
+    op.vmult_add(v,u);
+    v *= number;
+  };
+
+  return_op.Tvmult = [number, op](Domain &v, const Range &u)
+  {
+    op.Tvmult(v,u);
+    v *= number;
+  };
+
+  return_op.Tvmult_add = [number, op](Domain &v, const Range &u)
+  {
+    // FIXME: Is there a better way of doing this?
+    v /= number;
+    op.Tvmult_add(v,u);
+    v *= number;
+  };
+
+  return return_op;
+}
+
+
+/**
+ * Returns the transpose linear operations of @p op.
+ */
+template <typename Range, typename Domain>
+LinearOperator<Domain, Range>
+linop_transpose(const LinearOperator<Range, Domain> &op)
+{
+  LinearOperator<Domain, Range> return_op;
+
+  return_op.new_range_member = op.new_domain_member;
+  return_op.new_domain_member = op.new_range_member;
+
+  return_op.vmult = op.Tvmult;
+  return_op.vmult_add = op.Tvmult_add;
+  return_op.Tvmult = op.vmult;
+  return_op.Tvmult_add = op.vmult_add;
+
+  return return_op;
+}
+
+
+/**
+ * A function that encapsulates @p matrix objects that are compatible with
+ * dealii::Vector<Number> into a LinearOperator. The LinearOperator object
+ * that is created stores a reference to the matrix object. Thus, @p matrix
+ * must remain a valid reference for the whole lifetime of the
+ * LinearOperator object.
+ *
+ * All changes made on @p matrix after the creation of the LinearOperator
+ * object are reflected by the operator object. For example, it is a valid
+ * procedure to first create a LinearOperator and resize, reassemble the
+ * matrix later.
+ *
+ * The Matrix class in question must provide the following minimal
+ * interface:
+ *
+ * @code
+ * class Matrix
+ * {
+ * public:
+ *   // Dimension of domain space (matrix is of size $m\times n$)
+ *   dealii::Vector<Number>::size_type m() const;
+ *
+ *   // Dimension of range space (matrix is of size $m\times n$)
+ *   dealii::Vector<Number>::size_type n() const;
+ *
+ *   // Application of matrix to vector src. Writes result into dst
+ *   vmult(dealii::Vector<Number> &dst, const dealii::Vector<Number> &src);
+ *
+ *   // Application of transpose of matrix to vector src. Writes result
+ *   // into dst. (Depending on the usage of the linear operator class this
+ *   // can be a dummy implementation throwing an error)
+ *   Tvmult(dealii::Vector<Number> &dst, const dealii::Vector<Number> &src);
+ * };
+ * @endcode
+ *
+ * @authors: Matthias Maier, 2015
+ */
+template <template <typename> class Matrix, typename Number>
+LinearOperator<dealii::Vector<Number>, dealii::Vector<Number> >
+linop(const Matrix<Number> &matrix)
```

Comment:
All of our matrix classes have a local typedef `value_type`. I think it would be clearer if you replaced
```
template <template <typename> class Matrix, typename Number>
LinearOperator<dealii::Vector<Number>, dealii::Vector<Number> >
linop(const Matrix<Number> &matrix)
```
by
```
template <class Matrix>
LinearOperator<dealii::Vector<typename Matrix::value_type>, dealii::Vector<typename Matrix::value_type> >
linop(const Matrix &matrix)
```
because this avoids the (implicit) understanding that all matrix classes have a single template argument that denotes the scalar type.